translib/path/path_validator.go

```diff
@@ -0,0 +1,369 @@
+//  Unless required by applicable law or agreed to in writing, software       //
+//  distributed under the License is distributed on an "AS IS" BASIS,         //
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  //
+//  See the License for the specific language governing permissions and       //
+//  limitations under the License.                                            //
+//                                                                            //
+////////////////////////////////////////////////////////////////////////////////
+
+package path
+
+import (
+	"reflect"
+	"strings"
+
+	"fmt"
+
+	"github.com/Azure/sonic-mgmt-common/translib/ocbinds"
+	log "github.com/golang/glog"
+	"github.com/kylelemons/godebug/pretty"
+	"github.com/openconfig/gnmi/proto/gnmi"
+	"github.com/openconfig/goyang/pkg/yang"
+	"github.com/openconfig/ygot/util"
+	"github.com/openconfig/ygot/ygot"
+	"github.com/openconfig/ygot/ytypes"
+)
+
+type pathValidator struct {
+	gPath        *gnmi.Path
+	rootObj      *ocbinds.Device
+	sField       *reflect.StructField
+	sValIntf     interface{}
+	parentIntf   interface{}
+	opts         []PathValidatorOpt
+	parentSchema *yang.Entry
+	err          error
+}
+
+// NewPathValidator returns the PathValidator struct to validate the gnmi path and also add the missing module
+// prefix and key names and wild card values in the gnmi path based on the given PathValidatorOpt
+func NewPathValidator(opts ...PathValidatorOpt) *pathValidator {
+	return &pathValidator{rootObj: &(ocbinds.Device{}), opts: opts}
+}
+
+func (pv *pathValidator) init(gPath *gnmi.Path) {
+	gPath.Element = nil
+	pv.gPath = gPath
+	pv.sField = nil
+	pv.sValIntf = pv.rootObj
+	pv.parentIntf = pv.rootObj
+	pv.parentSchema = nil
+	pv.err = nil
+}
+
+// Validate the path and also add the module prefix / wild card keys in the gnmi path
+// based on the given PathValidatorOpt while creating the path validator.
+func (pv *pathValidator) Validate(gPath *gnmi.Path) error {
+	if pv == nil {
+		log.Warningf("error: pathValidator is nil")
+		return fmt.Errorf("pathValidator is nil")
+	}
+	if gPath == nil {
+		log.Warningf("error: gnmi path nil")
+		pv.err = fmt.Errorf("gnmi path nil")
+		return pv.err
+	}
+	if log.V(4) {
+		log.Info("Validate: path: ", gPath.Elem)
+	}
+	pv.init(gPath)
+	if pv.err = pv.validatePath(); pv.err != nil {
+		log.Warningf("error in validating the path: ", pv.err)
+	}
+	return pv.err
+}
+
+func (pv *pathValidator) getYangSchema() (*yang.Entry, error) {
+	sVal := reflect.ValueOf(pv.sValIntf)
+	if sVal.Elem().Type().Kind() == reflect.Struct {
+		objName := sVal.Elem().Type().Name()
+		if log.V(4) {
+			log.Info("getYangSchema: ygot object name: ", objName)
+		}
+		if ygSchema := ocbinds.SchemaTree[objName]; ygSchema == nil {
+			log.Warningf("error: ygot object name %v not found in the schema for the given path: %v", objName, pv.gPath)
+			return ygSchema, fmt.Errorf("invalid path %v", pv.gPath)
+		} else {
```

Comment:
Consider removing the else block since return occurs in the if block

Indent error flow: https://[google.github.io/styleguide/go/decisions#indent-error-flow
```
ygSchema := ocbinds.SchemaTree[objName]
if ygSchema == nil {
		log.Warningf("error: ygot object name %v not found in the schema for the given path: %v", objName, pv.gPath)
		return ygSchema, fmt.Errorf("invalid path %v", pv.gPath)
}
if log.V(4) {
		log.Infof("getYangSchema: found schema: %v for the field: %v", ygSchema.Name, *pv.sField)
}
return ygSchema, nil
```